File: benchmarks/cross-language/erlang/ping_pong.erl
-module(ping_pong).
-export([start/0, ping/3, pong/2]).

get_messages() ->
    case os:getenv("BENCHMARK_MESSAGES") of
        false -> 100000;
        Val -> list_to_integer(Val)
    end.

ping(Pong, N, _I) when N =< 0 ->
    Pong ! done,
    done;
ping(Pong, N, I) ->
    Pong ! {ping, self(), I},
    receive
        {pong, Value} ->
            % Validate received value matches what was sent
            if
                Value =/= I ->
                    io:format(standard_error, "Ping validation error: expected ~p, got ~p~n", [I, Value]);
                true -> ok
            end,
            ping(Pong, N - 1, I + 1)
    end.

pong(Expected, Messages) when Expected >= Messages ->
    receive
        done -> done
    end;
pong(Expected, Messages) ->
    receive
        {ping, Ping, Value} ->
            % Validate received value matches expected sequence
            if
                Value =/= Expected ->
                    io:format(standard_error, "Pong validation error: expected ~p, got ~p~n", [Expected, Value]);
                true -> ok
            end,
            % Echo back the EXACT value received
            Ping ! {pong, Value},
            pong(Expected + 1, Messages)
    end.

start() ->
    Messages = get_messages(),
    io:format("=== Erlang Ping-Pong Benchmark ===~n"),
    io:format("Messages: ~p~n~n", [Messages]),

    % Spawn processes
    Parent = self(),
    Pong = spawn(fun() -> pong(0, Messages) end),

    % Start timer (monotonic_time is correct for benchmarks per Erlang docs)
    Start = erlang:monotonic_time(nanosecond),

    % Run ping-pong - spawn returns the PID, we need to link it to receive done
    spawn_link(fun() ->
        ping(Pong, Messages, 0),
        Parent ! done
    end),

    % Wait for completion
    receive
        done -> ok
    after 60000 ->
        io:format("Timeout!~n"),
        halt(1)
    end,

    % Calculate metrics
    End = erlang:monotonic_time(nanosecond),
    ElapsedNs = End - Start,
    ElapsedSec = ElapsedNs / 1000000000.0,

    NsPerMsg = ElapsedNs / Messages,
    MsgPerSec = Messages / ElapsedSec,

    io:format("ns/msg:         ~.2f~n", [NsPerMsg]),
    io:format("Throughput:     ~.2f M msg/sec~n", [MsgPerSec / 1000000]),

    halt(0).

File: benchmarks/cross-language/elixir/ping_pong.exs
#!/usr/bin/env elixir

# Elixir ping-pong benchmark using processes
defmodule PingPong do
  @messages (System.get_env("BENCHMARK_MESSAGES") |> case do
    nil -> 100_000
    val -> String.to_integer(val)
  end)

  def ping(pong_pid, 0, _) do
    send(pong_pid, :done)
  end

  def ping(pong_pid, n, i) do
    send(pong_pid, {:ping, self(), i})
    receive do
      {:pong, value} ->
        # Validate received value matches what was sent
        if value != i do
          IO.puts(:stderr, "Ping validation error: expected #{i}, got #{value}")
        end
        ping(pong_pid, n - 1, i + 1)
    end
  end

  def pong(expected) do
    receive do
      {:ping, sender, value} ->
        # Validate received value matches expected sequence
        if value != expected do
          IO.puts(:stderr, "Pong validation error: expected #{expected}, got #{value}")
        end
        # Echo back the EXACT value received
        send(sender, {:pong, value})
        pong(expected + 1)
      :done ->
        :ok
    end
  end

  def run() do
    IO.puts("=== Elixir Ping-Pong Benchmark ===")
    IO.puts("Messages: #{@messages}")
    IO.puts("Using Erlang/OTP processes\n")

    ping_pid = self()
    pong_pid = spawn(fn -> pong(0) end)

    start = :erlang.monotonic_time(:nanosecond)
    ping(pong_pid, @messages, 0)
    finish = :erlang.monotonic_time(:nanosecond)

    total_ns = finish - start
    ns_per_msg = total_ns / @messages
    throughput = 1.0e9 / ns_per_msg

    IO.puts("ns/msg:         #{Float.round(ns_per_msg, 2)}")
    IO.puts("Throughput:     #{Float.round(throughput / 1.0e6, 2)} M msg/sec")
  end
end

PingPong.run()
